snuba/clickhouse/translator/visitor.py

```diff
@@ -0,0 +1,125 @@
+from copy import deepcopy
+from dataclasses import replace
+from typing import Optional, TypeVar
+
+from snuba.clickhouse.query import Expression as ClickhouseExpression
+from snuba.datasets.plans.translator.visitor import (
+    ExpressionMapper,
+    MappingExpressionTranslator,
+    TranslationRules,
+)
+from snuba.query.expressions import Expression as SnubaExpression
+from snuba.query.expressions import (
+    CurriedFunctionCall,
+    ExpressionVisitor,
+    FunctionCall,
+    Lambda,
+    SubscriptableReference,
+)
+
+
+class ExpressionTranslator(MappingExpressionTranslator[ClickhouseExpression]):
+    """
+    Translates a Snuba expression into a Clickhouse expression.
+
+    As long as the Clickhouse query AST is basically the same as the Snuba one
+    we add a default rule that makes a copy of the original expression.
+    """
+
+    def __init__(
+        self, translation_rules: TranslationRules[ClickhouseExpression]
+    ) -> None:
+        default_rules = TranslationRules(
+            literals=[DefaultSimpleMapper()],
+            columns=[DefaultSimpleMapper()],
+            # TODO: remove the DefaultSubscriptableMapper translation rule as
+            # soon as we finalize the tags translation rule.
+            subscriptables=[DefaultSubscriptableMapper()],
+            functions=[DefaultFunctionMapper()],
+            curried_functions=[DefaultCurriedFunctionMapper()],
+            arguments=[DefaultSimpleMapper()],
+            lambdas=[DefaultLambdaMapper()],
+        )
+        super().__init__(
+            translation_rules=translation_rules, default_rules=default_rules
+        )
+
+
+TSimpleExp = TypeVar("TSimpleExp", bound=SnubaExpression)
+
+
+class DefaultSimpleMapper(ExpressionMapper[TSimpleExp, ClickhouseExpression]):
+    def attemptMap(
+        self,
+        expression: TSimpleExp,
+        children_translator: ExpressionVisitor[ClickhouseExpression],
+    ) -> Optional[ClickhouseExpression]:
+        return ClickhouseExpression(deepcopy(expression))
```

Comment:
Something isn't adding up for me here.

[`ClickhouseExpression` is a new type derived from `snuba.query.expressions.Expression`](https://github.com/getsentry/snuba/blob/2b5cc5aea0b09fb647a61d176a7b95cfb8b994cc/snuba/clickhouse/query.py#L11-L13) (presumably because it will be very similar but represents a different domain so it makes sense to keep the class lineages separate) — but [`Expression` is an abstract base class](https://github.com/getsentry/snuba/blob/2b5cc5aea0b09fb647a61d176a7b95cfb8b994cc/snuba/query/expressions.py#L19-L64) that only has an `alias` constructor parameter, so instantiating `ClickhouseExpression` should throw, right? That is what `Expression` does when instantiated:

```
>>> from snuba.query.expressions import Expression
>>> Expression('asdf')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: Can't instantiate abstract class Expression with abstract methods __iter__, accept, transform
```

It seems like at runtime the type constructor becomes the identity function of the input (as if it were a cast), which explains why the tests are passing:

```python
>>> from snuba.clickhouse.translator.visitor import ClickhouseExpression
>>> o = object()
>>> ClickhouseExpression(o) is o
True
```

The runtime behavior is pretty clearly documented [in the `NewType` docs](https://docs.python.org/3/library/typing.html#newtype) — but it also says "The static type checker will treat the new type as if it were a subclass of the original type" which seems to be _not_ consistent with the abstract base class behavior.

For example, if this patch is applied to change `NewType` to a type alias instead (maintaining the _actual_ behavior of the constructor, which is something I'd expect that we want as soon as there is a real `ClickhouseExpression` type), like so:

```python
diff --git a/snuba/clickhouse/query.py b/snuba/clickhouse/query.py
index ab79432..c37034a 100644
--- a/snuba/clickhouse/query.py
+++ b/snuba/clickhouse/query.py
@@ -10,4 +10,4 @@ Query = NewType("Query", LogicalQuery)
 
 # This defines the type of a Clickhouse query expression. It is used to make
 # the interface of the expression translator more clear.
-Expression = NewType("Expression", SnubaExpression)
+Expression = SnubaExpression
```

Then mypy goes bonkers and everything in `tests/clickhouse/translator/test_visitor.py` fails with `TypeError: Can't instantiate abstract class Expression with abstract methods __iter__, accept, transform` (which is what I expected it to do in the first place):

```
snuba/clickhouse/translator/visitor.py:57: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:57: error: Argument 1 to "Expression" has incompatible type "TSimpleExp"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:66: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:67: error: Argument 1 to "Expression" has incompatible type "FunctionCall"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:84: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:85: error: Argument 1 to "Expression" has incompatible type "CurriedFunctionCall"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:105: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:106: error: Argument 1 to "Expression" has incompatible type "SubscriptableReference"; expected "Optional[str]"
snuba/clickhouse/translator/visitor.py:120: error: Cannot instantiate abstract class 'Expression' with abstract attributes '__iter__', 'accept' and 'transform'
snuba/clickhouse/translator/visitor.py:121: error: Argument 1 to "Expression" has incompatible type "Lambda"; expected "Optional[str]"
```

Is this something that you were aware of an intended to do? It seems like all of these would have to change to some extent as soon as an actual `ClickhouseExpression` type hierarchy exists — which I think will mostly affect this `DefaultSimpleMapper` since the output types will have to be the concrete types and not the abstract `ClickhouseExpression`.

Luckily, I do think that using `ClickhouseExpression` as the `TExpOut` for `ExpressionMapper` type generally makes sense (one exception in `DefaultCurriedFunctionMapper` noted below.)

Reply:
`Is this something that you were aware of and intended to do?`
Mostly. To me it is still an open point whether we will have a separate Expression hierarchy between Logical and Clickhouse. We have a reason now in the SubscriptableReference (which does not make sense in clickhouse) but that's it. So there are two options and I followed this approach. (declaring a type for clickhouseExpression) here to avoid getting stuck into a solution that would make it hard to introduce the clickhosue hierarchy.

1) We will have a separate hierarchy. This will have different types inside to an extent
In this case I don;t think there will be a common parent class that is not something above the already abstract ClickhouseExpression and SnubaExpression. In such case translation would not be a deepcopy. So the problem does not exist. We would be instantiating an object of a different type with its own constructor.

2) It will turn out we use the same expression hierarchy between logical and clickhouse
In this case I would argue we should not even have a type alias or a NewType. The translator will take an Expression and produce an Expression, since the hierarchy would be agnostic to whether it is part of the logical or the physical query. 
Again this has a problem in the SubscriptableReference, but, as we discussed today, there is an option in turning it into a FunctionCall. The question is how likely is it that we will be able to keep everything in one Expression hierarchy if/when we start doing some more advanced logical query processing.

3) ok, they were three. The two hierarchies look just the same but we want to keep different naming
I am not sure this would be better than 2 if there is no structural difference between the hierarchies, but in this case I would go with a plain type alias without trying to instantiate an abstract class.

I created the NewType here just to be sure we do not make implementation decisions that prevent us from picking 1 or 2 and leave us the time to figure out whether we actually want a separate Expression hierarchy for Clickhouse. I think it may be likely depending on how often we end up wanting new nodes in the Logical query to do logical query processing.
Anyway we cannot create a separate hierarchy as long as the Query class is still the same.